Explicitly generate an _external_ webhook URL for Withings integration

Flameeyes opened the pull request:
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

The current implementation of `webhook.async_generate_url()` will avoid
using the cloud URL, generating an internal URL that Withings will not
accept.

Force the URL to be generated disallowing internal URLs, and preferring
cloud instead.

Note: this is more of a short-term fix, since this is causing issues for users of 2022.6. A more proper solution would be to have a selection of internal/external when generating the webhook.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [X] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #72614
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

Flameeyes committed:
withings: attempt generating a cloud hook before falling back to webhook.

The current implementation of `webhook.async_generate_url()` will avoid
using the cloud URL, generating an internal URL that Withings will not
accept.

This uses the same implementation found in the toon component, trying to
set up a cloud hook first.

frenck commented on homeassistant/components/withings/common.py:
```diff
@@ -1030,6 +1030,11 @@ async def async_get_data_manager(
     if const.DATA_MANAGER not in config_entry_data:
         profile = config_entry.data.get(const.PROFILE)
 
+        webhook_url = "{}{}".format(
+            get_url(hass, allow_internal=False, prefer_cloud=True),
+            webhook.async_generate_path(config_entry.data[CONF_WEBHOOK_ID]),
+        )
```
We should not use `get_url` to generate cloud or webhooks.

For a reference implementation see:

<https://github.com/home-assistant/core/blob/79096864ebe6f187326289a3356341fea4252449/homeassistant/components/toon/coordinator.py#L58-L71>

Notes:
- The cloud hook needs to be stored and re-used
- Please use f-strings (instead of `format()` string interpolation).

Flameeyes replied:
Done! (Although I'd have expected this to be a bit more of a common requirement and something that would be happening more often.)

Unfortunately I don't have a setup where I can test this with Cloud enabled :/ any help would be welcome.

Flameeyes commented on homeassistant/components/withings/common.py:
```diff
@@ -1030,6 +1030,23 @@ async def async_get_data_manager(
     if const.DATA_MANAGER not in config_entry_data:
         profile = config_entry.data.get(const.PROFILE)
 
+        if CONF_CLOUDHOOK_URL not in config_entry_data:
```
What's a `.` vs a `_` between variables, huh? 😅

Thanks for catching that, should be fixed now, but it shows that I can't test this properly :/

frenck replied:
Not being able to test it, makes this kinda hard to accept as a PR, to be honest.
Maybe add unit tests to verify it's working instead?

Flameeyes replied:
The one instance where I can test withings is a friend's that has a publicly routable address, but not a Cloud subscription, that's why I couldn't test that behaviour easily.

I'll see what I can do for the tests — there's not much to look at that tests those cloudhook URLs but I guess I can borrow from the plaato config flow tests.

I mostly wanted to have something for those affected by #72614 since to fix one setup I broke the other 😣

mrlime replied:
How can I help test this?  I am affected by #72614 but I do not have a cloud subscription.  I access my Home Assistant via my own domain name and SSL certificate.

nebriv replied:
I have a cloud subscription if I can help test! Looks like I probably opened a duplicate issue here: https://github.com/home-assistant/core/issues/73200

Flameeyes replied:
Sorry folks, I've been a little [otherwise occupied](https://flameeyes.blog/2022/06/14/got-covid-it-sucks/) so I didn't pay enough attention here.

If you're comfortable with applying this change to your installation, you should be able to apply the diff cleanly on top of any 2022.6.x release, then restarting HA.

nebriv replied:
Ahh I got it sorted! The patch didn't have the changes mentioned here: 
https://github.com/home-assistant/core/pull/73228#discussion_r892421103

Manually fixing that resolves the KeyError. I am still getting the first MockRequest here though - it seems like withings is still making requests to the old webhook id (or maybe the incorrect endpoint, but still the correct webhook id?). I'm really not sure of the inner workings of the webhooks/home assistant configs yet.

Bluhme1 commented:
Any news on this? The work-around works ok but it is not a real solution.

nebriv commented:
@Flameeyes @vangorra 

Alright I think I got this working. In addition the patch above I also had to effectively comment out this line:
https://github.com/home-assistant/core/blob/dev/homeassistant/components/withings/__init__.py#L191

For some reason in the requests, maybe because its being proxied through nabu casa, the body_exists flag is not set.

These are the changes I made. Essentially just wrapping it in a try. I don't really know what the impact of doing this is though.
```python
    try:
        # Handle http post calls to the path.
        if not request.body_exists:
            return json_message_response("No request body", message_code=12)
    except AttributeError:
        _LOGGER.error("Caught Attribute Error, missing request_body")
        _LOGGER.debug(dir(request))
        # Not really sure what this is really checking here.
        # return json_message_response("No request body", message_code=12)
```

This results in the following log output if it useful to anyone:
```python
2022-06-15 20:43:21 ERROR (MainThread) [homeassistant.components.withings] Caught Attribute Error, missing request_body
2022-06-15 20:43:21 DEBUG (MainThread) [homeassistant.components.withings] ['__annotations__', '__class__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', '_content', '_text', 'content', 'headers', 'json', 'method', 'mock_source', 'post', 'query', 'query_string', 'status', 'text', 'url']
2022-06-15 20:43:21 DEBUG (MainThread) [homeassistant.components.withings] Withings webhook triggered
```

Flameeyes committed:
Implement MockRequest.body_exists to match aiohttp's Request.

This attribute is used in the withings integration, and is documented
as of aiohttp-2.3.

Flameeyes commented:
@nebriv thank you so much for testing this!

I've added a commit now that implements the `body_exists` attribute so that it matches the aiohttp class it's supposed to replace. Withings is the only webhook using it, so maybe removing the call (it does look superfluous to me) would be easier, but I'll leave that to @frenck to consider.

nathan-curtis commented:
Do we have an ETA on this? The workarounds isn't working for me on 2022.7 and no combination of base URL seems to be getting it back now. We've been waiting on this since February.

If you need help testing let us know what to test. But guys were stuck out here..

nebriv commented:
How long do these changes typically take to get approved and merged in? I'm kind of stuck not being able to update HomeAssistant until this patched. 😦 

I know there are a lot of pending review items - and appreciate the immense amount of work it takes triaging, reviewing, and implementing everything... especially for the low low price of free... But I'd love to see this one resolved!

Thanks all!

nebriv commented:
@frenck sorry for the ping - but how can we move this PR forward?

wez commented on homeassistant/components/withings/common.py:
```diff
@@ -1030,6 +1030,21 @@ async def async_get_data_manager(
+                data = {**config_entry.data, CONF_CLOUDHOOK_URL: webhook_url}
+                hass.config_entries.async_update_entry(config_entry, data=data)
+        else:
+            webhook_url = config_entry_data[CONF_CLOUDHOOK_URL]
```
I got a KeyError when trying to access `config_entry_data`, but it seems to me as though this whole if block is testing whether the URL is safe to access in `config_entry.data` and that it should be this instead:

```suggestion
            webhook_url = config_entry.data[CONF_CLOUDHOOK_URL]
```

wez replied:
In addition to this change, I found a number of steps were necessary to get this integration working:
https://github.com/home-assistant/core/issues/72614#issuecomment-1246218084

Importantly, subscriptions to the webhooks are always disabled AFAICT, and I needed to change the code for them to work at all:
https://github.com/home-assistant/core/issues/68760#issuecomment-1246206834

oneseventhree commented:
This is still an issue. Withings Bed Sensor only works if I manually put in the External URL to my Nabu Casa.

LCMilstein commented:
I am also still having trouble here.
I have 2 sensors. 1 of them is installed and works (I have to lay in bed after a restart), but it works.
The other, I removed (by accident) and I can NOT add the integration. It continues to report: {"errors":[{"message":"redirect_uri_mismatch: The redirect URI provided is missing or does not match partner callback url"}]}
Nothing is showing in the logs.
I've tried manually including the config in yaml (and then I get a warning that's been deprecated and I should remove it, but doesn't matter whether it's in or not). I've also tried manually modifying the const.py, manifest.json and common.py for the withings component... No change.

smileyhogue commented:
This is still an issue. I tried the work around that suggested using the generated webhook for the callback URL and that also did nothing. 

I am also getting the 293 error code.

pabloromeo commented:
The only way I was able to get the integration to work, was by following the steps here:
https://community.home-assistant.io/t/withings-integration-2022-4-2-redirect-uri-mismatch-error/410861/18

Basically manually re-enabled the "my:" integration in the config file, and then changing the callback URI in https://developer.withings.com/dashboard/ and setting the callback to https://my.home-assistant.io/redirect/oauth

smileyhogue commented:
For anyone using Nabu Casa, after 3 days of trial and error I have finally gotten it working. You seem to absolutely NOT be able to use your own domain at the moment. 
I had to change the external_url in configuration.yaml to use my nabucasa domain and after restarting, that worked for me. Though it still does fail occasionally after a restart of home assistant still.

frenck commented on homeassistant/components/withings/common.py:
```diff
@@ -1030,6 +1030,21 @@ async def async_get_data_manager(
+                data = {**config_entry.data, CONF_CLOUDHOOK_URL: webhook_url}
+                hass.config_entries.async_update_entry(config_entry, data=data)
+        else:
+            webhook_url = config_entry_data[CONF_CLOUDHOOK_URL]
```
```suggestion
            webhook_url = config_entry_data[CONF_CLOUDHOOK_URL]
```

☝️ This shows the code is untested.

frenck commented on homeassistant/components/withings/common.py:
```diff
@@ -1030,6 +1030,21 @@ async def async_get_data_manager(
     if const.DATA_MANAGER not in config_entry_data:
         profile = config_entry.data.get(const.PROFILE)
 
+        if CONF_CLOUDHOOK_URL not in config_entry.data:
```
We have to check if there is an active subscription, otherwise taking this path has no use

frenck reviewed (changes requested):
For this PR being able to merge:

- There is a merge conflict (minor)
- This PR is missing handling for not having a cloud subscription active.
- This PR is missing handling for people not having a cloud subscription (like not even a cloud component loaded at all, and use e.g., their own reverse proxies and things like that). 
- It needs to be tested (as mentioned in a way earlier review comment before as well).
- I'm going to require unit tests now as well, to show it works and ensure it keeps working as intended for now an in the future.

frenck commented:
I've marked this PR, as changes are requested that need to be processed.
Please un-draft it once it is ready for review again by clicking the **"Ready for review"** button.

Thanks! 👍 

../Frenck

[_Learn more about our pull request process._](https://developers.home-assistant.io/docs/review-process#prs-are-being-drafted-when-changes-are-needed)

GarethOates commented:
Any update on the status of this PR?  It seems like a lot of people are waiting for it to be resolved.

LCMilstein commented:
@frenck not sure if the webhook issue here is the only one, but confirming that Withings integration setup continues to fail for the in_bed binary sensor, reporting unavailable 100% of the time (not related to restarting HA, nor failing to initiate by lying in bed -- just fully will never come back with data to Home Assistant even though data is being logged into the Healthmate app).

johntdyer commented:
@Flameeyes  any change you can make the requested changes so we can get this one over the hump and into a release?   Thanks !!

joostlek commented:
As update, I hope to have Withings reworked in the next release, I already made major steps and this will be one of the features.

oneseventhree commented:
> As update, I hope to have Withings reworked in the next release, I already made major steps and this will be one of the features.

The birthday release or the one after :)?

joostlek commented:
hehe there is no birthday release. I am aiming for 2023.10

joostlek commented:
Will be fixed in #100916. Closing this one as it is too far behind at this point. Thanks for all the efforts ❤️
